### PR TITLE
Deprecate tuning enum and trainer properties

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Part of https://github.com/Lightning-AI/lightning/issues/11012
Addresses: https://github.com/Lightning-AI/lightning/pull/11089#discussion_r971871821

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @akihironitta @borda @rohitgr7 @tchaton

### CODE REVIEW

**Review by otaj — APPROVED** [review]

**Comment by rohitgr7** [comment]
> So many tests are failing...

yeah, forgot to remove the breakpoints.. 😅

**Review by justusschock — APPROVED** [review]

File: src/pytorch_lightning/trainer/states.py [hunk 1]
```diff
@@ -31,7 +59,7 @@ def stopped(self) -> bool:
         return self in (self.FINISHED, self.INTERRUPTED)
 
 
-class TrainerFn(LightningEnum):
+class TrainerFn(LightningEnum, metaclass=_DeprecatedEnumMeta):
```
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/states.py:62]:
```suggestion
class TrainerFn(LightningEnum, metaclass=_DeprecationManagingEnumMeta):
```

File: src/pytorch_lightning/trainer/states.py [hunk 2]
```diff
@@ -12,12 +12,40 @@
...
+from enum import Enum, EnumMeta
+from typing import Any, Optional
+
+from lightning_utilities.core.rank_zero import rank_zero_deprecation
 
 from pytorch_lightning.utilities import LightningEnum
 from pytorch_lightning.utilities.enums import _FaultTolerantMode
 
 
+class _DeprecatedEnumMeta(EnumMeta):
```
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/states.py:24]:
```suggestion
class _DeprecationManagingEnumMeta(EnumMeta):
```

File: src/pytorch_lightning/trainer/states.py [hunk 3]
```diff
@@ -44,16 +72,14 @@ class TrainerFn(LightningEnum):
...
+                f"`TrainerFn.{self.name}` has been deprecated in v1.8.0 and will be removed in v1.10.0."
+            )
 
-        This is used for the ``setup()`` and ``teardown()`` hooks
-        """
-        return TrainerFn.FITTING if self == TrainerFn.TUNING else self
 
-
-class RunningStage(LightningEnum):
+class RunningStage(LightningEnum, metaclass=_DeprecatedEnumMeta):
```
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/states.py:82]:
```suggestion
class RunningStage(LightningEnum, metaclass=_DeprecationManagingEnumMeta):
```

File: src/pytorch_lightning/trainer/states.py [hunk 4]
```diff
@@ -12,12 +12,40 @@
...
+
+    def __getattribute__(cls, name: str) -> Any:
+        obj = super().__getattribute__(name)
+        # ignore __dunder__ names -- prevents potential recursion errors
+        if not (name.startswith("__") and name.endswith("__")) and isinstance(obj, Enum):
+            obj.deprecate()
+        return obj
+
+    def __getitem__(cls, name: str) -> Any:
+        member: _DeprecatedEnumMeta = super().__getitem__(name)
```
Comment (otaj) [inline src/pytorch_lightning/trainer/states.py:38]:
```suggestion
        member: _DeprecationManagingEnumMeta = super().__getitem__(name)
```

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/strategies/strategy.py [hunk 5]
```diff
@@ -136,7 +136,7 @@ def setup_optimizers(self, trainer: "pl.Trainer") -> None:
         Args:
             trainer: the Trainer, these optimizers should be connected to
         """
-        if trainer.state.fn not in (TrainerFn.FITTING, TrainerFn.TUNING):
+        if trainer.state.fn != TrainerFn.FITTING:
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/strategy.py:139]: there's no way since the TrainerFn is overridden by their specific call.. fit/validate/test/predict.
I was going to propose to deprecate and remove `trainer.tune` in future since we have callbacks now but since `TrainerFn.TUNING` is redundant completely, I think deprecating it in this release would be fine.

File: src/pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -961,7 +961,7 @@ def _restore_modules_and_callbacks(self, checkpoint_path: Optional[_PATH] = None
     def _run(
         self, model: "pl.LightningModule", ckpt_path: Optional[str] = None
     ) -> Optional[Union[_EVALUATE_OUTPUT, _PREDICT_OUTPUT]]:
-        if self.state.fn in (TrainerFn.FITTING, TrainerFn.TUNING):
+        if self.state.fn == TrainerFn.FITTING:
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/trainer.py:964]: there's no way to determine tuning now

[the remark this answers is not in the document]
Comment (awaelchli) [reply src/pytorch_lightning/trainer/trainer.py:964]: That's problematic, because the logic here is guarding code that should not execute when optimizing parameters.

[the remark this answers is not in the document]
Comment (awaelchli) [reply src/pytorch_lightning/trainer/trainer.py:964]: I don't understand how it is possible to suddenly remove this logic just because of a deprecation?

[the remark this answers is not in the document]
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/trainer.py:964]: with the current logic, on the second call from within the callback, this function won't do anything since all of these values will be set when `.fit` is called.

File: src/pytorch_lightning/trainer/trainer.py [hunk 7]
```diff
@@ -1965,10 +1965,15 @@ def predicting(self, val: bool) -> None:
 
     @property
     def tuning(self) -> bool:
+        rank_zero_deprecation("`trainer.tuning` property has been deprecated in v1.8.0 and will be removed in v1.10.0.")
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/trainer.py:1968]:
```suggestion
        rank_zero_deprecation("`Trainer.tuning` has been deprecated in v1.8.0 and will be removed in v1.10.0.")
```
For consistency with our other messages

File: src/pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -1965,10 +1965,15 @@ def predicting(self, val: bool) -> None:
...
         return self.state.stage == RunningStage.TUNING
 
     @tuning.setter
     def tuning(self, val: bool) -> None:
+        rank_zero_deprecation(
+            "Setting `trainer.tuning` property has been deprecated in v1.8.0 and will be removed in v1.10.0."
+        )
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/trainer.py:1975]:
```suggestion
        rank_zero_deprecation(
            "Setting `Trainer.tuning` has been deprecated in v1.8.0 and will be removed in v1.10.0."
        )
```

File: tests/tests_pytorch/core/test_datamodules.py [hunk 9]
```diff
@@ -164,7 +164,6 @@ def test_train_loop_only(tmpdir):
     # fit model
     trainer.fit(model, datamodule=dm)
     assert trainer.state.finished, f"Training failed with {trainer.state}"
-    assert trainer.callback_metrics["train_loss"] < 1.0
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply tests/tests_pytorch/core/test_datamodules.py:167]: @rohitgr7 Can you check if it is related to your change `_SKLEARN_AVAILABLE = RequirementCache("scikit-learn")`. If so, please consider moving the fix to a separate PR

**Review by awaelchli — APPROVED** [review]

File: tests/tests_pytorch/trainer/connectors/test_checkpoint_connector.py [hunk 10]
```diff
@@ -123,22 +123,22 @@ def test_loops_restore(tmpdir):
...
-            trainer._checkpoint_connector.resume_start(ckpt_path)
-            trainer._checkpoint_connector.restore_loops()
-
-            trainer_loop = getattr(trainer, f"{fn}_loop")
-            trainer_loop.load_state_dict.assert_called()
-            trainer_loop.load_state_dict.reset_mock()
-
-        for fn2 in TrainerFn:
-            if fn2 not in (fn, TrainerFn.TUNING):
+    trainer_fns = [fn for fn in TrainerFn if fn != "tune"]
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/connectors/test_checkpoint_connector.py:126]: Can you introduce something like `TrainerFn._without_tune()` that returns this list? This way we remember when we end up removing it

**Review by carmocca — APPROVED** [review]
Thanks for following up!

File: src/pytorch_lightning/trainer/trainer.py (hunk 7 above)
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/trainer.py:1968]: done!

File: src/pytorch_lightning/trainer/trainer.py (hunk 8 above)
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/trainer.py:1975]: done!

File: tests/tests_pytorch/trainer/connectors/test_checkpoint_connector.py (hunk 10 above)
Comment (rohitgr7) [reply tests/tests_pytorch/trainer/connectors/test_checkpoint_connector.py:126]: done!

File: tests/tests_pytorch/core/test_datamodules.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (rohitgr7) [reply tests/tests_pytorch/core/test_datamodules.py:167]: let me try..

[the remark this answers is not in the document]
Comment (rohitgr7) [reply tests/tests_pytorch/core/test_datamodules.py]: seems like the case.. weird.. sklearn is a package.. was failing for me locally.
